client/components/update-post-status/index.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { flow, once } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import UpdateTemplate from './update-template';
+import PostActions from 'lib/posts/actions';
+import { recordGoogleEvent } from 'state/analytics/actions';
+
+const RESET_TIMEOUT_MS = 1200;
+
+const getStrings = once( ( translate ) => ( {
+	page: {
+		deleteWarning: translate( 'Delete this page permanently?' ),
+		deleted: translate( 'Page Deleted' ),
+		deleting: translate( 'Deleting Page' ),
+		error: translate( 'Error' ),
+		restored: translate( 'Page Restored' ),
+		restoring: translate( 'Restoring Page' ),
+		trashed: translate( 'Moved to Trash' ),
+		trashing: translate( 'Trashing Page' ),
+		undo: translate( 'undo?' ),
+		updated: translate( 'Updated' ),
+		updating: translate( 'Updating Page' ),
+	},
+	post: {
+		deleteWarning: translate( 'Delete this post permanently?' ),
+		deleted: translate( 'Post Deleted' ),
+		deleting: translate( 'Deleting Post' ),
+		error: translate( 'Error' ),
+		restored: translate( 'Restored' ),
+		restoring: translate( 'Restoring' ),
+		trashed: translate( 'Moved to Trash' ),
+		trashing: translate( 'Trashing Post' ),
+		undo: translate( 'undo?' ),
+		updated: translate( 'Updated' ),
+		updating: translate( 'Updating Post' ),
+	},
+} ) );
+
+const enhance = flow(
+	localize,
+	connect( null, { recordGoogleEvent } )
+);
+
+const updatePostStatus = ( WrappedComponent ) => enhance(
+	class UpdatePostStatus extends Component {
+		static displayName = `UpdatePostStatus(${
+			WrappedComponent.displayName || WrappedComponent.name || ''
+		})`
+
+		static propTypes = {
+			translate: PropTypes.func.isRequired,
+			recordGoogleEvent: PropTypes.func.isRequired,
+			post: PropTypes.object,
+			page: PropTypes.object,
+		}
+
+		state = {
+			updated: false,
+			updatedStatus: null,
+			previousStatus: null,
+			showMoreOptions: false,
+			showPageActions: false,
+		}
+
+		getType() {
+			return this.props.page ? 'page' : 'post';
+		}
+
+		buildUpdateTemplate = () => {
+			if ( ! this.state.updated ) {
+				return;
+			}
+
+			const strings = getStrings( this.props.translate );
```

Comment:
Not the most elegant, but I'm convinced it will do for now. Trying not to waste too much time on the HoC and focus on getting rid of the mixin.